### PR TITLE
Convert interconnect udp tests from tinc to regress

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
All existing interconnect udp tests are converted from tinc to regress,
they need about 16 minutes to finish, so we put them in a new test
target installcheck-icudp to prevent slowing ICW.

On pipeline we also made a bit changes to the running environment, we
used to run the tests in a multiple segments cluster, now we run
directly on a demo cluster.

### CODE REVIEW

**Comment by pivotal-ning-yu** [comment]
A dev pipeline is setup for this PR, it's green:
https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/dev-icudp?groups=Interconnect

**Comment by ashwinstar** [comment]
This is great to see, thank you.

Seems only one test takes 15 mins rest are very small
```
test icudp/icudp_full         ... ok (954.80 sec)  (diff:0.12 sec)
```
wondering what that big test is.

Also, can we possibly add the smaller tests to regular ICW, just keep the big one separate? (question just to know dependency and all, depends we really do it or not)

**Comment by pivotal-ning-yu** [comment]
A commit was pushed to move the fast icudp tests into ICG.

**Comment by ashwinstar** [comment]
Might be worth looking (even if not now as part of this PR) are the 15 mins spent in that test really worth spent. Seems some pretty big unions and select repeats in there like

```sql
SELECT jkey2, SUM(length(digits_string)) AS sum_len_dstring
  FROM (
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
    UNION ALL
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
    UNION ALL
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
    UNION ALL
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
    UNION ALL
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
    UNION ALL
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
    UNION ALL
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
    UNION ALL
    (SELECT jkey % 30 AS jkey2, repeat('0123456789', 200) AS digits_string FROM small_table GROUP BY jkey2)
UNION ALL
...........
...
```

I do not have much context, so can't advice on how to cut it down. But seems should be able to test similar stuff with much reduced time. 15 full ICW runs currently, so single test taking that much time in parallel seems still pretty high.

**Comment by pivotal-ning-yu** [comment]
Made some experiments, looks like the slowest query is below one:

```sql
-- Huge tuple (May need to split) 26 * 200000
SELECT SUM(length(long_tval)) AS sum_len_tval
  FROM (SELECT jkey, repeat(tval, 200000) AS long_tval
          FROM small_table ORDER BY dkey LIMIT 20) foo
            JOIN (SELECT * FROM small_table ORDER BY dkey LIMIT 500) bar USING(jkey);
```

The execution time are listed below when setting different GUCs:
- 6.6s by default;
- 89s when `gp_udpic_fault_inject_percent = 40` and `gp_udpic_fault_inject_bitmap = 50790655`;
- 148s when `gp_udpic_dropxmit_percent = 20`;
- 87s when `gp_udpic_dropacks_percent = 20`;
- 407s when setting all above;

But I do not know how to explain these timings.  @pengzhout do you have any idea?

**Comment by pengzhout** [comment]
Those GUC control the drop rate of interconnect packets like DATA/ACK/STOP/EOS type packets, the higher the drop rate is, the longer time interconnect will take to resend those packets. Consider we run ICW with high frequency, I'm OK with setting it from 40 to 20 or even 15, so we can reduce enough time to put it to ICW tests.

File: src/test/regress/expected/icudp/icudp_full.out [hunk 1]
```diff
@@ -744,6 +744,10 @@ RESET gp_udpic_dropxmit_percent;
 RESET gp_udpic_fault_inject_percent;
```
Comment (ashwinstar) [inline src/test/regress/expected/icudp/icudp_full.out]: Don't think RESET's as on session level only, drop table and drop schema required.

Comment (pivotal-ning-yu) [reply src/test/regress/expected/icudp/icudp_full.out]: Makes sense, a commit was pushed to remove the useless RESET commands.

**Comment by pivotal-ning-yu** [comment]
Pushed a commit to run a low drop rate version of icudp_full in ICW, the execution time for it is about 270s.  We still kept the installcheck-icudp target, the original high drop rate version of icudp_full still run here.

The other fast icudp tests are now run in both installcheck-icudp and ICW, the idea is to make installcheck-icudp a centered place for all the icudp tests.

**Comment by pivotal-ning-yu** [comment]
No more reviews for 4 days, I will merge it on master now, thanks for the review.

**Comment by ashwinstar** [comment]
Interconnect tests are consistently failing without asserts due to
```
--- /tmp/build/e18b2f02/gpdb_src/src/test/regress/expected/icudp/icudp_full.out	2018-10-23 21:42:12.142211547 +0000
+++ /tmp/build/e18b2f02/gpdb_src/src/test/regress/results/icudp/icudp_full.out	2018-10-23 21:42:12.162211547 +0000
@@ -18,10 +18,14 @@

-- Makesure these codes are triggered to improve the code coverage during test.
 SET gp_udpic_dropacks_percent = 20;
+ERROR:  unrecognized configuration parameter "gp_udpic_dropacks_percent"
 SET gp_udpic_dropxmit_percent = 20;
+ERROR:  unrecognized configuration parameter "gp_udpic_dropxmit_percent"
 SET gp_udpic_fault_inject_percent = 40;
+ERROR:  unrecognized configuration parameter "gp_udpic_fault_inject_percent"
 SET gp_interconnect_full_crc = true;
 SET gp_udpic_fault_inject_bitmap = 50790655;
+ERROR:  unrecognized configuration parameter "gp_udpic_fault_inject_bitmap"
 -- SET gp_log_interconnect TO 'DEBUG';
 -- SET gp_interconnect_log_stats = true;
```

Those GUCs are under `#ifdef USE_ASSERT_CHECKING` so seems the test is not relevant at all for without asserts. Should we really have tests relevant only for asserts build or is there way to code them differently maybe not using guc to enable running without asserts as well?

**Comment by pivotal-ning-yu** [comment]
@ashwinstar Thanks for the reminder.  These gucs simulate packets dropping which are only used for debugging & testing and have negative effect on the performance slightly.  We are working on finding out some way to prevent this job being generated on the without_assert pipeline, there is a internal story to track this.

**Pull request merged**